### PR TITLE
feat: Added equality-based and set-based filtering by label keys and values; for list, sync and wait App commands.

### PR DESCRIPTION
**Pull request by maheshbaliga**
This PR only resolves the `-l / --selector` flag part of the issue, where apps can be filtered with theirs labels and values. It achieves feature-parity with [kubectl](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/). 

List, Sync and Wait commands to manage applications can now be filtered by label keys and values. Matching apps must satisfy all of the specified label constraints, though they may have additional labels as well. Equality-based operators allowed are =, == and !=. Set-based operators allowed are exists, not exists, in and notin.

Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [ ] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [x] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

File: server/application/application_test.go [hunk 1]
```diff
@@ -301,6 +301,104 @@ func createTestApp(testApp string, opts ...func(app *appsv1.Application)) *appsv
...
+			label: "key1<value1"},
+	}
+	//test invalid scenarios
+	for _, invalidTest := range invalidTests {
+		t.Run(invalidTest.testName, func(t *testing.T) {
+			appQuery.Selector = &invalidTest.label
+			res, err := appServer.List(context.Background(), &appQuery)
+			assert.Error(t, err)
+			assert.Nil(t, res)
+		})
```
Comment (ApsTomar) [inline server/application/application_test.go:398]: Since, these are invalid cases so we know that they're going to throw some error like "invalid operator". So, we need to match the error with that exact error. Just one assert statement will do.

**Review by ApsTomar — CHANGES_REQUESTED** [review]

File: docs/user-guide/commands/argocd_app_list.md [hunk 2]
```diff
@@ -24,7 +24,7 @@ argocd app list [flags]
   -o, --output string          Output format. One of: wide|name|json|yaml (default "wide")
   -p, --project stringArray    Filter by project name
   -r, --repo string            List apps by source repo URL
-  -l, --selector string        List apps by label
+  -l, --selector string        List apps by label. Supports equality-based (=, ==, !=) and set-based (in, notin, exists and not exists) operators.
```
Comment (samkulkarni20) [inline docs/user-guide/commands/argocd_app_list.md:27]: Same message can be copy-pasted here.

File: docs/user-guide/commands/argocd_app_sync.md [hunk 3]
```diff
@@ -48,7 +48,7 @@ argocd app sync [APPNAME... | -l selector | --project project-name] [flags]
       --retry-backoff-max-duration duration   Max retry backoff duration. Input needs to be a duration (e.g. 2m, 1h) (default 3m0s)
       --retry-limit int                       Max number of allowed sync retries
       --revision string                       Sync to a specific revision. Preserves parameter overrides
-  -l, --selector string                       Sync apps that match this label
+  -l, --selector string                       Sync apps that match this label. Supports equality-based (=, ==, !=) and set-based (in, notin, exists and not exists) operators.
```
Comment (samkulkarni20) [inline docs/user-guide/commands/argocd_app_sync.md:51]: Same as above

File: docs/user-guide/commands/argocd_app_wait.md [hunk 4]
```diff
@@ -27,7 +27,7 @@ argocd app wait [APPNAME.. | -l selector] [flags]
   -h, --help                   help for wait
       --operation              Wait for pending operations
       --resource stringArray   Sync only specific resources as GROUP:KIND:NAME. Fields may be blank. This option may be specified repeatedly
-  -l, --selector string        Wait for apps by label
+  -l, --selector string        Wait for apps by label. Supports equality-based (=, ==, !=) and set-based (in, notin, exists and not exists) operators.
```
Comment (samkulkarni20) [inline docs/user-guide/commands/argocd_app_wait.md:30]: +1

File: cmd/argocd/commands/app.go [hunk 5]
```diff
@@ -1725,7 +1725,7 @@ func NewApplicationSyncCommand(clientOpts *argocdclient.ClientOptions) *cobra.Co
 	command.Flags().BoolVar(&prune, "prune", false, "Allow deleting unexpected resources")
 	command.Flags().StringVar(&revision, "revision", "", "Sync to a specific revision. Preserves parameter overrides")
 	command.Flags().StringArrayVar(&resources, "resource", []string{}, fmt.Sprintf("Sync only specific resources as GROUP%sKIND%sNAME. Fields may be blank. This option may be specified repeatedly", resourceFieldDelimiter, resourceFieldDelimiter))
-	command.Flags().StringVarP(&selector, "selector", "l", "", "Sync apps that match this label")
+	command.Flags().StringVarP(&selector, "selector", "l", "", "Sync apps that match this label. Supports equality-based (=, ==, !=) and set-based (in, notin, exists and not exists) operators.")
```
Comment (samkulkarni20) [inline cmd/argocd/commands/app.go:1728]: In all three help texts here, the help message sounds too technical. They should be in layman language instead. This is what kubectl help says for label selectors:
```
    -l, --selector='':
	Selector (label query) to filter on, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2). Matching
	objects must satisfy all of the specified label constraints.
```

File: server/application/application_test.go (hunk 1 above)
Comment (maheshbaliga) [reply server/application/application_test.go:400]: Done.

File: docs/user-guide/commands/argocd_app_list.md (hunk 2 above)
Comment (maheshbaliga) [reply docs/user-guide/commands/argocd_app_list.md]: Done.

File: docs/user-guide/commands/argocd_app_sync.md (hunk 3 above)
Comment (maheshbaliga) [reply docs/user-guide/commands/argocd_app_sync.md]: Done.

File: docs/user-guide/commands/argocd_app_wait.md (hunk 4 above)
Comment (maheshbaliga) [reply docs/user-guide/commands/argocd_app_wait.md]: Done.

File: cmd/argocd/commands/app.go (hunk 5 above)
Comment (maheshbaliga) [reply cmd/argocd/commands/app.go]: Done.
